src/shared/conf-parser.c

```diff
@@ -568,21 +590,127 @@ int config_parse_many(
                 void *userdata,
                 Hashmap **ret_stats_by_path) {
 
-        _cleanup_strv_free_ char **dropin_dirs = NULL;
         _cleanup_strv_free_ char **files = NULL;
-        const char *suffix;
         int r;
 
-        suffix = strjoina("/", dropin_dirname);
-        r = strv_extend_strv_concat(&dropin_dirs, (char**) conf_file_dirs, suffix);
+        assert(conf_file_dirs);
+        assert(dropin_dirname);
+        assert(sections);
+        assert(table);
+
+        r = config_get_dropin_files(conf_file_dirs, dropin_dirname, &files);
+        if (r < 0)
+                return r;
+
+        return config_parse_many_files(conf_files, files, sections, lookup, table, flags, userdata, ret_stats_by_path);
+}
+
+static int config_get_stats_by_path_one(
+                const char* conf_file,
+                const char* const* conf_file_dirs,
+                Hashmap *stats_by_path) {
+
+        _cleanup_strv_free_ char **files = NULL;
+        _cleanup_free_ char *dropin_dirname = NULL;
+        struct stat st;
+        int r;
+
+        assert(conf_file);
+        assert(conf_file_dirs);
+        assert(stats_by_path);
+
+        /* Unlike config_parse(), this does not support stream. */
+
+        r = path_extract_filename(conf_file, &dropin_dirname);
         if (r < 0)
                 return r;
+        if (r == O_DIRECTORY)
+                return -EINVAL;
+
+        if (!strextend(&dropin_dirname, ".d"))
+                return -ENOMEM;
 
-        r = conf_files_list_strv(&files, ".conf", NULL, 0, (const char* const*) dropin_dirs);
+        r = config_get_dropin_files(conf_file_dirs, dropin_dirname, &files);
         if (r < 0)
                 return r;
 
-        return config_parse_many_files(conf_files, files, sections, lookup, table, flags, userdata, ret_stats_by_path);
+        /* First read the main config file. */
+        r = RET_NERRNO(stat(conf_file, &st));
+        if (r >= 0) {
+                r = hashmap_put_stats_by_path(&stats_by_path, conf_file, &st);
+                if (r < 0)
+                        return r;
+        } else if (r != -ENOENT)
+                return r;
+
+
+        /* Then read all the drop-ins. */
+        STRV_FOREACH(fn, files) {
+                if (stat(*fn, &st) < 0) {
+                        if (errno == ENOENT)
+                                continue;
+
+                        return -errno;
+                }
+
+                r = hashmap_put_stats_by_path(&stats_by_path, *fn, &st);
+                if (r < 0)
+                        return r;
+        }
+
```

Comment:
this should also detect mtime of the drop-in dir itself, so that we can detect added files.

Reply:
`stats_by_path_equal` does a hashmap size comparison which should detect newly added and deleted files including the drop-ins?

Reply:
Here, I think @gogsbread is right. storing stats for drop-in dir is not necessary, and moreover, it triggers unnecessary reload: e.g. if we have 
- /usr/lib/systemd/network/10-eth0.link.d/a.conf
- /usr/lib/systemd/network/10-eth0.link.d/b.conf
- /etc/systemd/network/10-eth0.link.d/a.conf

Then, it is not necessary to reload configs when /usr/lib/systemd/network/10-eth0.link.d/a.conf is updated, but we need to reload configs when /usr/lib/systemd/network/10-eth0.link.d/b.conf is updated.

Reply:
I can't think of a scenario where the current approach wouldn't detect added files either.

Reply:
One question: Do we store full paths in the stats_by_path hashmap or just the relative path to /usr or /etc or some other config directory? Asking because if we store the full path, then updating /usr/lib/systemd/network/a.conf if /etc/systemd/network/a.conf exists will lead to a reload no, even if it isn't needed in that case because the file in /etc overrides the file in /usr?

Reply:
If /usr/lib/systemd/network/10-eth0.link.d/a.conf exists, then /etc/systemd/network/10-eth0.link.d/a.conf is not listed by `config_get_dropin_files()`. Then, storing full path does not trigger unnecessary reload.

Moreover, when /usr/lib/systemd/network/10-eth0.link.d/a.conf is removed, and /etc/systemd/network/10-eth0.link.d/a.conf has take an effect, then we should reload configs. Of course, in that case, `struct stat` is also different, but it is better to detect the change also from the path. Hence, I think it is better to store full path.